### PR TITLE
Fix issue147

### PR DESCRIPTION
**Pull request by pscff**
Fix build-issues on Windows. Tests failing: 24-55.

### CODE REVIEW

File: doc/build-with-vs.md [hunk 1]
```diff
@@ -1,77 +1,72 @@
...
+5. Build from the Windows commandline
+
+   $ mkdir .\out\build\x64-Release-vs15
+   $ cd .\out\build\x64-Release-vs15
+   $ cmake -G "Visual Studio 15 2017" ^
+           -A x64 ^
+           -DWITH_DOC=OFF ^
+           -DWITH_HDF5=OFF ^
+           -DCMAKE_CONFIGURATION_TYPES:STRING="Release" ^
+           -DCMAKE_INSTALL_PREFIX:PATH="%CD%\..\..\install\x64-Release" ^
```
Comment (jesper-friis) [inline doc/build-with-vs.md:36]: %CD% is not recognised on my system

**Review by jesper-friis — COMMENTED** [review]
This is very good - I was actually able to compile and run some of the tests.

All tests in src/utils/tests works 
All Python tests works
All tests in src/tests fails for me

That the Python tests works is great!
That the C tests in src/tests fails is probably due to bad environment variables. Have you set anything system-wise?

File: src/utils/CMakeLists.txt [hunk 2]
```diff
@@ -44,10 +45,17 @@ check_symbol_exists(stat                sys/stat.h   HAVE_STAT)
 check_symbol_exists(exec                unistd.h     HAVE_EXEC)
 check_symbol_exists(P_tmpdir            stdio.h      HAVE_P_TMPDIR)
 
+if(HAVE_WINDOWS_H)
+check_symbol_exists(GetFullPathNameW    windows.h    HAVE_GetFullPathNameW)
+check_symbol_exists(GetFullPathName     windows.h    HAVE_GetFullPathName)
+check_symbol_exists(GetFileAttributesW  windows.h    HAVE_GetFileAttributesW)
+check_symbol_exists(GetFileAttributes   windows.h    HAVE_GetFileAttributes)
+else()
```
Comment (jesper-friis) [inline src/utils/CMakeLists.txt:53]: I guess that line 53-57 can be removed completely since it works on mingw to include `windows.h` as well and POSIX systems neither have these functions or fileapi.h

**Review by jesper-friis — APPROVED** [review]
Very nice. Good description of how to build. For me tests 24-40 and 49-55 fails, but the Python tests goes through.

I have added some minor comments, but the commit is good.

File: doc/build-with-vs.md (hunk 1 above)
Comment (jesper-friis) [reply doc/build-with-vs.md:36]: Sorry, this is fine in the Windows prompt. I was just git bash...

File: doc/build-with-vs.md [hunk 3]
```diff
@@ -1,77 +1,72 @@
...
-    PATH=$PATH:$LOCAL_DIR/bin
-    cd dlite
-    mkdir build && cd build
-    cmake -G "Visual Studio 14 2015 Win64" -DHDF5_DIR=$LOCAL_DIR/cmake/hdf5 -DCMAKE_INSTALL_PREFIX=$LOCAL_DIR/ ..
-    cmake --build . --config Debug --target doc
-    cmake --build . --config Debug --target install
-    cmake --build . --config Release --target install
-    ctest -C Debug
-    ctest -C Release
+1. Install Visual Studio 15 2017 or Visual Studio 16 2019 with the
```
[the remark this answers is not in the document]
Comment (pscff) [reply doc/build-with-vs.md:4]: OK, even better. I will update the documentation accordingly.

[the remark this answers is not in the document]
Comment (pscff) [reply doc/build-with-vs.md:4]: I like it when tests pass, but in this case it puzzles me why for your Python-tests work. The PATH environment is not set correctly via CMake (See https://github.com/SINTEF/dlite/issues/149#issue-1003926905) - this is because semicolon path-separators are interpreted as a list in CMakes set_property(TEST §{test} ENVIRONMENT "PATH=${dlite_PATH_NATIVE}"). The python package can therefore not find the dlls via the PATH. But where does it find your dlls? The AddDllDirectory workaround in ./bindings/python/__init__.py is not working on my system. Is it working on yours? or do you have dlite.dll and dlite-utils.dll somewhere Windows look for dlls by default (Such as the system-directory?) maybe from a previous buildt?

[the remark this answers is not in the document]
Comment (jesper-friis) [reply doc/build-with-vs.md:4]: I have no idea why it works for me but not for you. I use git bash together with the below script to configure, compile and run the tests. Maybe the ways dlite handle environment variables plays better with bash than the dos prompt? But I have tried to make sure that all paths confirms to Windows... 

```bash
#!/bin/sh

set -xe

export SOURCE_ROOT="C:\Users\jesperf\source\repos\dlite"
export PYTHON_LIBDIR="C:\Program Files\Python38"


# Compile
export BUILD_ROOT="${SOURCE_ROOT}\build-win64"
mkdir -p "${BUILD_ROOT}"
cd ${BUILD_ROOT}
cmake -G "Visual Studio 15 2017 Win64" \
      -A x64 \
      -DWITH_DOC=OFF \
      -DWITH_HDF5=OFF \
      -DCMAKE_CONFIGURATION_TYPES:STRING="Release" \
      -DCMAKE_INSTALL_PREFIX:PATH="${SOURCE_ROOT}\install\win64" \
      ..

cmake --build . --config Release
cmake --install .
ctest -C Release


# Test with build root
export DLITE_USE_BUILD_ROOT=YES
export PATH="${BUILD_ROOT}\src\Release;${BUILD_ROOT}\src\utils\Release;${PYTHON_LIBDIR}"
export PYTHONPATH="${BUILD_ROOT}\bindings\python"
export DLITE_STORAGES="${SOURCE_ROOT}/src/tests/*.json"

#src/tests/Release/test_storage.exe
python.exe "${SOURCE_ROOT}\bindings\python\tests\test_python_bindings.py"



# Test installed
export INSTALL_PREFIX="${SOURCE_ROOT}\install\win64"

export DLITE_USE_BUILD_ROOT=NO
export DLITE_ROOT="${INSTALL_PREFIX}"
export PATH="${INSTALL_PREFIX}\bin;${PYTHON_LIBDIR}"
export PYTHONPATH="${INSTALL_PREFIX}/lib/site-packages"

#src/tests/Release/test_storage.exe
#python.exe "${SOURCE_ROOT}\bindings\python\tests\test_python_bindings.py"
```

[the remark this answers is not in the document]
Comment (pscff) [reply doc/build-with-vs.md:4]: I tried this in Git bash. "ctest -C Release" is still failing for tests 41-48 on my system. The dlls cannot be found. Since I am using Anaconda, I might be hitting the following issue: https://github.com/pytorch/pytorch/issues/17051. If that is the case, the AddDllDirectory mechanism in ./bindings/python/\_\_init\_\_.py is not working for me but it may be working for your system with a different Python distribution.

When I set the PATH via export etc. the Python tests pass both in Git bash or from the Windows command prompt.

Summing up, I believe it is best to remove the AddDllDirectory components from \_\_init\_\_.py and set the PATH correctly in CMake (https://github.com/SINTEF/dlite/issues/149) so that the tests can pass. For now, can we close this issue and merge into master?
